BJS2-15563

Caveri95 opened the pull request:
- Создал класс контроллер с методами для работы с событиями (регистрация/отмена регистрации/получения списка участников/количество участников)
- Создал класс сервис с логикой по работе со слоем репозитория событий
- Добавил тесты

Caveri95 committed:
BJS2-15563

- Создал класс контроллер с методами для работы с событиями (регистрация/отмена регистрации/получения списка участников/количество участников)
- Создал класс сервис с логикой по работе со слоем репозитория событий
- Добавил тесты

ravshanbekn commented on src/main/java/school/faang/user_service/dto/EventDto.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Min(1)
+    @Max(128)
+    private String location;
```
По моему аннотации @ Min @ Max нужны для численных типов
Для String:
@ Size(min = 1, message = "validation message")
@ Size(max = 64, message = "validation message")
Еще я бы указал сообщение в message аргумент

Caveri95 replied:
Да, точно, ты прав, что то забыл про нее. В Size сразу можно указать два аргумента на нижнюю и верхнюю границы

Caveri95 replied:
Если провалиться в message, то там неплохие такие заготовки на описание ошибки

ravshanbekn commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,28 @@
+    @NotBlank
+    @Min(3)
+    @Max(64)
+    private String username;
```
Тут тоже через аннотацию @ Size

ravshanbekn reviewed (commented):
Думаю, в целом, нормально

Caveri95 committed:
BJS2-15563

- Поправил аннотацию на валидность
- Создал класс на проверку наличия участника на ивенте/убрал метод из репозитория на поиск таких участников
- Убрал интерфейс на сервис

pablo023 commented on src/main/java/school/faang/user_service/service/eventService/EventParticipationServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+    private final ParticipantOnEventValidator participantOnEventValidator;
+
+    public void registerParticipant(long eventId, long userId) {
+        if (!participantOnEventValidator.checkParticipantAtEvent(eventId, userId)) {
+            eventParticipationRepository.register(eventId, userId);
+        } else {
+            throw new IllegalArgumentException("User with id " + userId + " is already registered for event with id " + eventId);
+        }
+    }
```
Тут есть валидация, её лучше в отдельный класс - какой-то, в духе.. EventParticipantsValidator. И чтобы просто метод начинася примерно с такой строчки:

eventParticipantsValidator.validate(eventId, userId);

Над неймингом, конечно, подумай, чтобы название отображало суть валидации конкретной. Можно конечно сделать через перегруженный метод, чтобы был один validate для разных сценариев, но это опсано - может получиться как лучше, так и хуже, надо как следует подумать.

Caveri95 replied:
Вынес валидацию в отдельный класс, но по ощущениям так стало только сложнее.
Если метод валидации делаю войд, то тестировать его уже не очень становится и в методе сервиса тест будет только на то, что у нас дернется валидация и, например, регистрация. Хотя регистрация может и не вызваться, если валидация не пройдет.
В самом классе валидации есть два метода, по сути раньше они отличались только возвращаемым значением  true/false, сейчас они войд и приходится разделять эти методы. 
В общем посмотри пожалуйста.

i7rincpane replied:
Да необычно, что сценарии разные, может тогда лучше чтобы каждый сценарий свой валидатор имел и сделать типо, как с фильтром, динамическое добавление ХХХ)))))) насчет тестов, если правильно понял, вроде как verifyNoInteractions можно вызвать проверить, что регистрация и мапер не дернулись.. но вообще юнит тесты тоже смущают сильно, сам себе создаешь какие-то ситуации, и не понятно проверяешь ли вообще что то? поскорей бы интеграционные, там как то все более очевидно кажется

pablo023 replied:
> Да необычно, что сценарии разные, может тогда лучше чтобы каждый сценарий свой валидатор имел и сделать типо, как с фильтром, динамическое добавление ХХХ)))))) насчет тестов, если правильно понял, вроде как verifyNoInteractions можно вызвать проверить, что регистрация и мапер не дернулись.. но вообще юнит тесты тоже смущают сильно, сам себе создаешь какие-то ситуации, и не понятно проверяешь ли вообще что то? поскорей бы интеграционные, там как то все более очевидно кажется

Такова природа юнит-тестов. Если тебя это смущает - обрати внимание на TDD. Мне такой подход неоднократно помогал начать что-то писать, когда я не понимал - с какой стороны подступиться, и юнит-тесты обретают больше смысла.

pablo023 commented on src/main/java/school/faang/user_service/service/eventService/EventParticipationServiceImpl.java:
Всю валидацию надо вынести в отдельный класс, и его методы.

pablo023 reviewed (changes requested):
Очень хорошо) Немного исправь, и едем дальше!

Caveri95 committed:
BJS2-15563

- Перенес методы валидации в класс для валидации
- Добавил тесты на валидацию
- Поменял путь к папкам для тестов

Caveri95 committed:
BJS2-15563

- Поменял путь к папкам для тестов

pablo023 commented on src/main/java/school/faang/user_service/Validate/EventParticipantValidator.java:
Если тебя смущают тут войды, ты можешь сделать их булеанами. В целом войды не должны смущать, тестить их не так удобно, но научиться можно. Пакет переименуй на validator, форс пуш, и смерджи сам

pablo023 reviewed (approved):
Молодец)

pablo023 commented:
И конфликты пофиксь.

Caveri95 committed:
BJS2-15563

- Изменил название пакета на validator
